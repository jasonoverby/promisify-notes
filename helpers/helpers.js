const chalk = require('chalk');

const notANumber = num => typeof num !== 'number' || Number.isNaN(num);
const getWaitedSecsMsg = num => `waited for ${num} seconds`;
const getNotANumberMsg = (...strings) => (
  strings.length === 1
    ? `"${strings[0]}" is not a number`
    : `"One or more of these arguments is not a number: ${strings.join(', ')}`
);
const getImprovedErrorMsg = err => (
  chalk.red(err.stack.toString().split('\n').slice(0, 2).join('\n'))
);

const getRandomIntBetweenMinAndMax = (num1, num2) => {
  // ensures non-negative integers
  const [abs1, abs2] = [num1, num2].map(Math.abs).map(Math.floor);
  const [min, max] = [Math.min(abs1, abs2), Math.max(abs1, abs2)];
  return Math.floor((Math.random() * (max - min)) + min);
};

module.exports = {
  getImprovedErrorMsg,
  getNotANumberMsg,
  getRandomIntBetweenMinAndMax,
  getWaitedSecsMsg,
  notANumber,
};
