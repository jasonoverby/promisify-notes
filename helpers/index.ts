import chalk from 'chalk';

const notANumber = (num: any): boolean => typeof num !== 'number' || Number.isNaN(num);
const getWaitedSecsMsg = (num: number): string => `waited for ${num} second(s)`;
const getNotANumberMsg = (num1: number, num2: number): string => (
  `"One or both of these arguments is not a number: ${num1}, ${num2}`);
const getFancyString = (str: string) => chalk.magenta(`${str} fancy`);
const patt = new RegExp(/waited for \d second\(s\)/);
const fancyPatt = new RegExp(/waited for \d second\(s\) for something fancy/);

const getRandomIntBetweenMinAndMax = (num1: number, num2: number): number => {
  // ensures non-negative integers
  const [abs1, abs2] = [num1, num2].map(Math.abs).map(Math.floor);
  const [min, max] = [Math.min(abs1, abs2), Math.max(abs1, abs2)];
  return Math.floor((Math.random() * (max - min)) + min);
};

export {
  fancyPatt,
  getFancyString,
  getNotANumberMsg,
  getRandomIntBetweenMinAndMax,
  getWaitedSecsMsg,
  notANumber,
  patt,
};
