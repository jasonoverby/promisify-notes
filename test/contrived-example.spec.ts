import { promisify } from 'util';
import { patt } from '../helpers';
import {
  getMsgAfterRandomSecsWithCallback,
  NodeStyleCallback,
} from '../lib/contrived-example';

describe('promisifyContrivedExample', () => {
  const num1 = 1;
  const num2 = 3;
  const badInput = 'a';
  const getMsgAfterRandomSecsAsync = promisify(
    getMsgAfterRandomSecsWithCallback,
  );

  beforeEach(() => {
    jest.restoreAllMocks();
    jest.spyOn(global, 'setTimeout');
  });

  describe('callback pattern', () => {
    it('gets a message after a random number of seconds inside a callback', (done) => {
      const theCallback: NodeStyleCallback = (_: Error, fancyMsg: string) => {
        expect(fancyMsg).toMatch(patt);
        done();
      };

      getMsgAfterRandomSecsWithCallback(num1, num2, theCallback);
      expect(setTimeout).toHaveBeenCalledTimes(1);
      expect.assertions(2);
    });

    it('gets an error when one of the args is not a number using a callback', (done) => {
      const theCallback: NodeStyleCallback = (err: Error) => {
        expect(err).toBeInstanceOf(TypeError);
        expect(err).toMatchSnapshot();
        done();
      };

      // @ts-ignore bad input on purpose
      getMsgAfterRandomSecsWithCallback(badInput, num2, theCallback);
      expect(setTimeout).toHaveBeenCalledTimes(1);
      expect(setTimeout).toHaveBeenCalledWith(expect.any(Function), 0);
      expect.assertions(4);
    });
  });

  describe('async/await pattern', () => {
    it('gets a message after a random number of seconds using the promisified function', async () => {
      const fancyMsg = await getMsgAfterRandomSecsAsync(num1, num2);

      expect(fancyMsg).toMatch(patt);
      expect(setTimeout).toHaveBeenCalledTimes(1);
      expect.assertions(2);
    });

    it('throws an error when one of the args is not a number using the promisified function', async () => {
      await expect(
        // @ts-ignore bad input on purpose
        getMsgAfterRandomSecsAsync(badInput, num2),
      ).rejects.toThrowError(TypeError);

      expect(setTimeout).toHaveBeenCalledTimes(1);
      expect(setTimeout).toHaveBeenCalledWith(expect.any(Function), 0);
      expect.assertions(3);
    });
  });
});
