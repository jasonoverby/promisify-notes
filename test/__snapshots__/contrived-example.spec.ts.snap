// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`promisifyContrivedExample async/await pattern gets an error after 0 seconds when one of the args is not a number using the promisified function 1`] = `
"[31mTypeError: \\"One or both of these arguments is not a number: a, 3[39m
[31m    at Timeout._onTimeout (/Users/j0o00um/p/practice/promisify/lib/contrived-example.ts:20:16)[39m"
`;

exports[`promisifyContrivedExample callback pattern gets an error after 0 seconds when one of the args is not a number using a callback 1`] = `[TypeError: "One or both of these arguments is not a number: a, 3]`;
